sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala

```diff
@@ -109,28 +110,75 @@ object ScalaNativePluginInternal {
         .withDump(nativeDump.value)
     },
     nativeLink := {
+      val classpath = fullClasspath.value.map(_.data.toPath)
       val outpath = (nativeLink / artifactPath).value
-      val config = {
-        val mainClass = selectMainClass.value.getOrElse {
-          throw new MessageOnlyException("No main class detected.")
+
+      def makeBuild(): HashFileInfo = {
+        val config = {
+          val mainClass = selectMainClass.value.getOrElse {
+            throw new MessageOnlyException("No main class detected.")
+          }
+
+          val maincls = mainClass + "$"
+          val cwd = nativeWorkdir.value.toPath
+
+          val logger = streams.value.log.toLogger
+          build.Config.empty
+            .withLogger(logger)
+            .withMainClass(maincls)
+            .withClassPath(classpath)
+            .withWorkdir(cwd)
+            .withCompilerConfig(nativeConfig.value)
+        }
+
+        interceptBuildException {
+          Build.build(config, outpath.toPath)(sharedScope)
         }
-        val classpath = fullClasspath.value.map(_.data.toPath)
-        val maincls = mainClass + "$"
-        val cwd = nativeWorkdir.value.toPath
-
-        val logger = streams.value.log.toLogger
-        build.Config.empty
-          .withLogger(logger)
-          .withMainClass(maincls)
-          .withClassPath(classpath)
-          .withWorkdir(cwd)
-          .withCompilerConfig(nativeConfig.value)
+
+        FileInfo.hash(outpath)
       }
 
-      interceptBuildException {
-        Build.build(config, outpath.toPath)(sharedScope)
+      def buildIfChanged(): Unit = {
+        import sbt.util.CacheImplicits._
+        import collection.JavaConverters._
+
+        val cacheFactory = streams.value.cacheStoreFactory / "fileInfo"
+        val classpathTracker =
+          Tracked.inputChanged[Seq[HashFileInfo], HashFileInfo](
+            cacheFactory.make("inputFileInfo")
+          ) {
+            case (changed: Boolean, filesInfo: Seq[HashFileInfo]) =>
+              val outputTracker =
+                Tracked
+                  .lastOutput[Seq[HashFileInfo], HashFileInfo](
+                    cacheFactory.make("outputFileInfo")
+                  ) {
+                    case (_, None) => makeBuild()
+                    case (_, Some(outHashInfo)) =>
+                      if (changed || outHashInfo != FileInfo.hash(outpath))
+                        makeBuild()
+                      else outHashInfo
+                  }
```

Comment:
This probably can be also squished to only two branches, which can be helpful if we would move hashing out of makeBuild() body
```suggestion
                  ) { (_, prev) => 
                   val outputHashInfo = FileInfo.hash(outpath) 
                   if(changed || !prev.contains(outputHashInfo)) {
                     makeBuild()
                   }
                   outputHashInfo
                  }
```